fix: use correct path length derivatives when computing the curvilinear covariance for a zero step propagation

goetzgaycken opened the pull request:
The propagate can use curvilinear parametrisation for the returned parameters and covariance, However if the propagation step size is below the limit, the propagator does not compute the path length derivatives which leads to an incorrect covariance matrix when enabling curvilinear parameterization.

paulgessinger commented:
Thanks @goetzgaycken! This would have to be rebased on top of `main` itself, instead of `v31.2.0`, which should also resolve the conflict, I believe.

paulgessinger commented:
@beomki-yeo and @andiwand could you maybe have a look at this?

goetzgaycken commented on Core/include/Acts/Propagator/Propagator.ipp:
```diff
@@ -256,6 +284,36 @@ auto Acts::Propagator<S, N>::makeResult(propagator_state_t state,
   moveStateToResult(state, result);
 
   if (makeCurvilinear) {
+    if constexpr (has_extension<decltype(state.stepping)>::value &&
```
Not clear whether this is sufficient to test whether the enclosed lines are valid

goetzgaycken commented on Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp:
```diff
@@ -0,0 +1,487 @@
+
+}  // namespace
+
+struct TestData {
```
The test data does not contain the expected output for the covariance matrix in curvilinear parametrisation. It is assumed that the propagator correctly computes it for at least one step size or that the alternative method is correct. However, both methods call the same methods. So, while for one acts version it was correct, the test would not be sensitive to all kinds of changes.

goetzgaycken replied:
There does not seem to be a strong opinion. closing.

andiwand commented on Core/include/Acts/Propagator/Propagator.ipp:
```diff
@@ -256,6 +284,36 @@ auto Acts::Propagator<S, N>::makeResult(propagator_state_t state,
   if (makeCurvilinear) {
+    if constexpr (has_extension<decltype(state.stepping)>::value &&
+                  has_stepData<decltype(state.stepping)>::value) {
+      if (state.stepping.pathAccumulated <= 0.) {
```
right, this can be negative with back propagation

andiwand commented on Core/include/Acts/Propagator/Propagator.ipp:
```diff
@@ -256,6 +284,36 @@ auto Acts::Propagator<S, N>::makeResult(propagator_state_t state,
+          auto fieldRes = m_stepper.getField(state.stepping, pos);
+          if (fieldRes.ok()) {
+            sd.B_first = *fieldRes;
+            if (state.stepping.extension.k1(state, m_stepper, m_navigator,
+                                            sd.k1, sd.B_first, sd.kQoP)) {
+              // dr/ds :
+              state.stepping.derivative.template head<3>() =
+                  state.stepping.pars.template segment<3>(eFreeDir0);
+              // d (dr/ds) / ds :
+              state.stepping.derivative.template segment<3>(4) = sd.k1;
+              // to set dt/ds :
+              state.stepping.extension.finalize(state, m_stepper, m_navigator,
+                                                state.stepping.pathAccumulated);
+            }
+          }
```
do you think this can be done in the eigen stepper instead?

that way we can also avoid the template based checks if this applies or not

goetzgaycken replied:
You are right, It should go into the EigenStepper. Similar modifications would be needed in the AtlasStepper. It is less clear to me what is needed for the StraightLineStepper. I suppose the StraightLineStepper should produce the same as the EigenStepper in the limit B->0. However in this limit only the d (dr/ds) / ds are zero not the other components of the path length derivatives, and I do not see that being accounted for.

andiwand commented on Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp:
```diff
@@ -0,0 +1,487 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#define HAS_DERIV
```
can this be dropped?

andiwand reviewed (commented):
for my understanding: is the missing term from the covariance depending on the magnetic field? because I would have thought the contribution from the direction should be 0 since we always arrive perpendicular on the curvilinear surface

goetzgaycken commented:
> for my understanding: is the missing term from the covariance depending on the magnetic field? 

It depends on both the direction but also the cross product of the magnetic field and the time components on the propagation velocity. Even without a magnetic field a small correction results from the path length derivatives.

goetzgaycken commented:
changes:  f9938ddbc34ad7b8af5e392481f3076bd71ae9f6..8b4961907 :
- implemented @andiwand 's comment and moved the derivative computation to the stepper. 
- the unit test now compares the curvilinear covariance matrices of the explicit computation with small step propagation using eignstepper, and straightline stepper, where for the latter the magnetic field of the test data is forced to zero. 

Propagation using the AtlasStepper and /MultiEigenStepperLoop have not been tested and/or fixed.

andiwand commented on Core/include/Acts/Propagator/Propagator.ipp:
```diff
@@ -256,6 +256,10 @@ auto Acts::Propagator<S, N>::makeResult(propagator_state_t state,
   moveStateToResult(state, result);
 
   if (makeCurvilinear) {
+    if (!m_stepper.prepareCurvilinearState(state, m_navigator)) {
+      // information to compute curvilinearState is incomplete.
+      return propagationResult.error();
+    }
```
do you think this can be triggered within the stepper? for example during state creation

goetzgaycken replied:
> do you think this can be triggered within the stepper? for example during state creation

The functions called by the eigen stepper implementation could fail in theory e.g. if the extensions are not valid for the "current step", and I think the computation of the magnetic field can potentially fail. In practice it is unlikely.

andiwand replied:
Is it possible to implement this purely inside the stepper without touching the propagator?

to me it looks a bit odd that we have to touch the propagator for this as it looks like a stepper issue to me.

but maybe I am overlooking something

goetzgaycken replied:
Yes, but then curvilinearState would have to become a template, which presumably would mean to add one layer on-top of curvilinearState  e.g.:
```
    template< class T_Navigator, class T_PropagatorState>
    Stepper::curvilinearState(T_PropagatorState prop_state, T_Navigator navigator nav) {
      if (!prepareCurvilinearState(prop_state, nav)) {
        return; // error
      }
      return curvilinearState_impl(prop_state.stepping);
    }
```

andiwand replied:
I see. not sure if we could wire these templates

alternatively we could change the propagator to make a 0 distance step even if we are on target to begin with

any other suggestions @paulgessinger @asalzburger ?

beomki-yeo commented:
The PR looks good to me as long as the comments from others are resolved!

andiwand reviewed (dismissed):
I left a few more comments and suggestions.

Generally I feel like there should be an easier solution to this but I also don't have a concrete idea for that so no objection from my side to merge this as it is since it fixes a problem.

beomki-yeo commented on Core/include/Acts/Propagator/EigenStepper.ipp:
```diff
@@ -57,6 +57,43 @@ auto Acts::EigenStepper<E, A>::boundState(
+              prop_state.stepping.pars.template segment<3>(eFreeDir0);
+          // d (dr/ds) / ds :
+          prop_state.stepping.derivative.template segment<3>(4) = sd.k1;
+          // to set dt/ds :
```
Its not going to.
Dqop/ds requires the material information

beomki-yeo replied:
And its zero for vacuum so it can be skipped for this setup I guess

beomki-yeo replied:
For dense yes. 
I thought this code is in the test directory - but it's in the eigen stepper.

andiwand replied:
this should be handled in the extension https://github.com/acts-project/acts/blob/main/Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp

goetzgaycken replied:
Thanks !

goetzgaycken replied:
The implementation prepareCurvilinearState will call extension.finalize which should compute d qop / ds if the dense environment extension is used. The unit test does not cover dense materials and the reference path length derivatives which are used in that test are only valid in the vacuum.

goetzgaycken commented on Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp:
```diff
@@ -0,0 +1,602 @@
+  Vector3 bfield;
+};
+
+template <typename T_StepperCreator>
+void test_bound_to_curvilinear(const std::vector<TestData> &test_data_list,
+                               const T_StepperCreator &stepper_creator) {
```
I don't think it really matters for the test.

Anyway the EigenStepper does not allow to swap out the MagneticFieldProvider, and each test data item provides a constant magnetic field for which a new provider is needed. I could implement a special magnetic field provider which provides a special magnetic field for each area which concerns the test data item, but that would be more complicated.

goetzgaycken commented on Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp:
```diff
@@ -0,0 +1,602 @@
+std::vector<TestData> make_test_data(double mag_field_scale = 1.) {
+  std::vector<TestData> test_data_list{
+      TestData(
+          makeVector<ActsScalar, 3>(
+              {-442.883, -624.094, 857.272}),  // surface center
+          makeMatrix<ActsScalar, 3, 3>(
+              {0.677197, 0.0176111, -0.735591, -0.735342, -0.0191232, -0.677426,
+               -0.0259971, 0.999662, -2.22045e-16}),  // surface rot
```
Eigen only supports initializer_list constructors starting from a certain version (I forgot which one), and one ubuntu test, if I recall correctly, uses a slightly too old Eigen version. It also  only worked for vectors and not matrices. I implemented the initializer_list helper functions, to make the test data definition a bit more concise. I think even without looking at the implementation of these helpers it is clear what the result is supposed to be. Although for matrices it is never clear whether the data needs to be given in column or row major order, but this does not matter for covariance matrices. Anyway, I would argue that it helps with the readability.

goetzgaycken commented on Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp:
```diff
@@ -0,0 +1,602 @@
+    const Vector3 &direction, double qop, const Vector3 &bfield,
+    const T_ParticleHypothesis &particle_hypothis) {
+  FreeToPathMatrix path_length_deriv;
+#if defined(EIGEN_HAS_CONSTEXPR) && EIGEN_VERSION_AT_LEAST(3, 4, 0)
+  static_assert(path_length_deriv.cols() ==
+                8);  // ensure that all elements are initialized
+#endif
```
Yes, up to the moment somebody decides to add extra degrees of freedom (I suppose the time component for example was not there from the very beginning, or am I wrong ?)

goetzgaycken commented on Tests/UnitTests/Core/Propagator/BoundToCurvilinearConversionTests.cpp:
```diff
@@ -0,0 +1,602 @@
+  return false;
+}
+
+#define MSG_DEBUG(a)               \
+  if (isDebugOutputEnabled()) {    \
+    std::stringstream msg;         \
+    msg << a;                      \
+    BOOST_TEST_MESSAGE(msg.str()); \
+  }                                \
+  do {                             \
+  } while (0)
```
Since it is a boost test, I thought that it makes more sense to use the "logging mechanism" provided by the boost test framework. In this way all the command line arguments should have the desired behavior.

paulgessinger commented on Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:
```diff
@@ -759,6 +759,21 @@ class MultiEigenStepperLoop
+  /// @param [in] navigator the navigator of the propagation
+  /// @return true if nothing is missing after this call, false otherwise.
+  template <typename propagator_state_t, typename navigator_t>
+  bool prepareCurvilinearState(
```
Does this not have the same issue? Does the correction call somehow get forwarded to the `EigenStepper`, and I'm not seeing it?

goetzgaycken commented on Core/include/Acts/Propagator/EigenStepper.hpp:
```diff
@@ -339,6 +339,18 @@ class EigenStepper {
+  /// @param [in] navigator the navigator of the propagation
+  /// @return true if nothing is missing after this call, false otherwise.
+  template <typename propagator_state_t, typename navigator_t>
+  bool prepareCurvilinearState(propagator_state_t& prop_state,
```
In principle yes, though the has_method helper template does not seem to support arguments which are templates.

andiwand commented:
@goetzgaycken there is a `clang-tidy` failure https://github.com/acts-project/acts/pull/2910/checks?check_run_id=21308679491 can you fix that?

goetzgaycken commented:
update from 5ec58b2567ba0506913dd058e51610dbf32b6ec3  to 97a96a877db67b16897dc2861ff24b3026913815
rebase to main of  2024-03-04 12:29
fixed :
> @goetzgaycken there is a `clang-tidy` failure https://github.com/acts-project/acts/pull/2910/checks?check_run_id=21308679491 can you fix that?

goetzgaycken commented:
There seems to be a problem in "CI Bridge / build_linux_ubuntu".  Though I do not see a particular what goes wrong exactly. Is this just an infrastructure problem ?

paulgessinger commented:
I think this is an OOM. I'm retrying.

acts-project-service commented:
## :red_circle: Athena integration test results [8329f20541a0730fc34168f7657afa88c1a85867]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37288999) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7090962)!
